sklearn/semi_supervised/self_training.py

```diff
@@ -0,0 +1,239 @@
+import numpy as np
+
+from ..base import BaseEstimator, clone
+from ..utils.validation import check_X_y, check_array, check_is_fitted
+from ..utils.metaestimators import if_delegate_has_method
+from ..utils import safe_mask
+
+__all__ = ["SelfTrainingClassifier"]
+
+# Authors: Oliver Rausch   <rauscho@ethz.ch>
+#          Patrice Becker  <beckerp@ethz.ch>
+# License: BSD 3 clause
+
+
+def _check_estimator(estimator):
+    """Make sure that an estimator implements the necessary methods."""
+
+    if not hasattr(estimator, "predict_proba"):
+        raise ValueError("The base_estimator should implement predict_proba!")
+
+
+class SelfTrainingClassifier(BaseEstimator):
+    """Self-training classifier
+
+    Parameters
+    ----------
+    base_estimator : estimator object
+        An estimator object implementing ``fit`` and ``predict_proba``.
+        Invoking the ``fit`` method will fit a clone of the passed estimator,
+        which will be stored in the ``self.base_estimator_`` attribute.
+
+    threshold : float
+        Threshold above which predictions are added to the labeled dataset.
+        Should be in [0, 1).
+
+    max_iter : integer
+        Maximum number of iterations allowed. Should be greater than or equal
+        to 0.
+
+    Attributes
+    ----------
+    base_estimator_: estimator object
+        The fitted estimator.
+
+    y_labeled_ : array, shape = (n_samples)
+        The labels assigned to unlabeled datapoints during fitting.
+
+    X_labeled_ : array, shape = (n_samples, n_features)
```

Comment:
If we can avoid storing something potentially as big as this, we should do so. We can store the index into X instead...?

Reply:
In the new implementation of fit, this isn't stored anymore because y_labeled_ and y_labeled_iter_ now are of length n_samples (They were previously of length n_labeled_samples), and therefore the indicies of y_labeled_ match the indicies of the passed X.